src/irr/asset/CGLTFLoader.cpp

```diff
@@ -387,26 +411,46 @@ namespace irr
 
 					auto getShaders = [&](bool hasUV, bool hasColor) -> std::pair<core::smart_refctd_ptr<ICPUSpecializedShader>, core::smart_refctd_ptr<ICPUSpecializedShader>>
 					{
-						auto loadShaders = [&](const std::string_view& filename) -> core::smart_refctd_ptr<ICPUSpecializedShader>
+						auto loadShader = [&](const std::string_view& cacheKey) -> core::smart_refctd_ptr<ICPUSpecializedShader>
 						{
-							auto shader_bundle = assetManager->getAsset("include/irr/builtin/shaders/loaders/gltf/" + std::string(filename.data()), {});
-
-							bool status = !shader_bundle.isEmpty();
-							assert(status);
-
-							auto shader = core::smart_refctd_ptr_static_cast<ICPUSpecializedShader>(shader_bundle.getContents().begin()[0]);
-							return shader;
+							size_t storageSz = 1ull;
+							asset::SAssetBundle bundle;
+							const IAsset::E_TYPE types[]{ IAsset::ET_SPECIALIZED_SHADER, static_cast<IAsset::E_TYPE>(0u) };
+
+							assetManager->findAssets(storageSz, &bundle, cacheKey.data(), types);
+							if (bundle.isEmpty())
+								return nullptr;
+							auto assets = bundle.getContents();
+							
+							return core::smart_refctd_ptr_static_cast<ICPUSpecializedShader>(assets.begin()[0]);
 						};
 
 						if (hasUV) // if both UV and Color defined - we use the UV
-							return std::make_pair(loadShaders(SShaderPaths::UV::vertex), loadShaders(SShaderPaths::UV::fragment));
+							return std::make_pair(loadShader(VERT_SHADER_UV_CACHE_KEY), loadShader(FRAG_SHADER_UV_CACHE_KEY));
 						else if (hasColor)
-							return std::make_pair(loadShaders(SShaderPaths::Color::vertex), loadShaders(SShaderPaths::Color::fragment));
+							return std::make_pair(loadShader(VERT_SHADER_COLOR_CACHE_KEY), loadShader(FRAG_SHADER_COLOR_CACHE_KEY));
 						else
-							return std::make_pair(loadShaders(SShaderPaths::NoUVAndColor::vertex), loadShaders(SShaderPaths::NoUVAndColor::fragment));
+							return std::make_pair(loadShader(VERT_SHADER_NO_UV_COLOR_CACHE_KEY), loadShader(VERT_SHADER_NO_UV_COLOR_CACHE_KEY));
```

Comment:
you should have entire pipelines pre-loaded and pre-cached

Reply:
not just shaders, just like MTL

Reply:
jo, **todo**!

Reply:
wait, but the loader need to return meshes not pipelines, so to make pipeline cashed I need to insert them here manually to a cache or create another gltf pipeline loader that will be returning a bundle of pipelines (cached) and I need a way to associate them with a mesh being rendered

Reply:
what should I do?

Reply:
I dont mean that glTF loader should be like MTL loader in everything...

I mean that there's a finite amount of combinations of shaders and pipeline parameters, and you can have them as "built-in assets"

Basically turn all the options (vertex inputs, etc.) into a unique string for a `builtin` glTF pipeline, then instead of creating a new pipeline every single time a glTF is loaded, do this:
1. If pipeline with this unique string (hash) is found in asset cache, then
1a. If its not a dummy object then use it
1b. If its a dummy object but you don't need to inspect it (get any of the info that would be garbage for a dummy) then use it
1c. If its a dummy, and that would be a problem, recreate the pipeline and then restore the dummy from the created one (needs a PR merge from @Crisspl on IrrlichtBaW)

P.S. it would make sense to do the same to your `builtin` shaders as your builtin pipelines.

Reply:
so I will create an interface for transmission those options into string and then will be following your hints

Reply:
As I begun creating it I think it isn't a good idea or you should give me more details 

`Basically turn all the options (vertex inputs, etc.) into a unique string for a builtin glTF pipeline, then instead of creating a new pipeline every single time a glTF is loaded, do this:`

we will be in troubles as soon as somebody wants to query the loaded asset due to translation so I stopped working further with the interface and I'm waiting for details.